frontend/lib/src/components/widgets/Slider/Slider.tsx

```diff
@@ -308,133 +169,408 @@ class Slider extends React.PureComponent<Props, State> {
+    dataType === SliderProto.DataType.DATE ||
+    dataType === SliderProto.DataType.TIME
+  )
+}
+
+function formatValue(value: number, element: SliderProto): string {
+  const { format, options } = element
+  if (isDateTimeType(element)) {
+    // Python datetime uses microseconds, but JS & Moment uses milliseconds
+    // The timestamp is always set to the UTC timezone, even so, the actual timezone
+    // for this timestamp in the backend could be different.
+    // However, the frontend component does not need to know about the actual timezone.
+    return moment.utc(value / 1000).format(format)
+  }
+
+  if (options.length > 0) {
+    return sprintf(format, options[value])
+  }
+
+  return sprintf(format, value)
+}
+
+/**
+ * Return the value of the slider. This will either be an array with
+ * one value (for a single value slider), or an array with two
+ * values (for a range slider).
+ */
+function getValueAsArray(value: number[], element: SliderProto): number[] {
+  const { min, max } = element
+  let start = value[0]
+  let end = value.length > 1 ? value[1] : value[0]
+  // Adjust the value if it's out of bounds.
+  if (start > end) {
+    start = end
+  }
+  if (start < min) {
+    start = min
+  }
+  if (start > max) {
+    start = max
+  }
+  if (end < min) {
+    end = min
+  }
+  if (end > max) {
+    end = max
+  }
+  return value.length > 1 ? [start, end] : [start]
+}
+
+function fixLabelPositions(
+  sliderDiv: HTMLDivElement | null,
+  thumb1Div: HTMLDivElement | null,
+  thumb2Div: HTMLDivElement | null,
+  thumb1ValueDiv: HTMLDivElement | null,
+  thumb2ValueDiv: HTMLDivElement | null
+): void {
+  if (!sliderDiv || !thumb1Div || !thumb1ValueDiv) {
+    return
+  }
+
+  fixLabelOverflow(sliderDiv, thumb1Div, thumb1ValueDiv)
+
+  if (thumb2Div && thumb2ValueDiv) {
+    fixLabelOverflow(sliderDiv, thumb2Div, thumb2ValueDiv)
+
+    // If two thumbs.
+    fixLabelOverlap(
+      sliderDiv,
+      thumb1Div,
+      thumb2Div,
+      thumb1ValueDiv,
+      thumb2ValueDiv
     )
   }
 }
 
-export default withTheme(Slider)
+function fixLabelOverflow(
+  slider: HTMLDivElement,
+  thumb: HTMLDivElement,
+  thumbValue: HTMLDivElement
+): void {
+  const sliderRect = slider.getBoundingClientRect()
+  const thumbRect = thumb.getBoundingClientRect()
+  const thumbValueRect = thumbValue.getBoundingClientRect()
+
+  const thumbMidpoint = thumbRect.left + thumbRect.width / 2
+  const thumbValueOverflowsLeft =
+    thumbMidpoint - thumbValueRect.width / 2 < sliderRect.left
+  const thumbValueOverflowsRight =
+    thumbMidpoint + thumbValueRect.width / 2 > sliderRect.right
+
+  thumbValue.style.left = thumbValueOverflowsLeft ? "0" : ""
+  thumbValue.style.right = thumbValueOverflowsRight ? "0" : ""
+}
+
+/**
+ * Goals:
+ * - Keep the thumb values near their respective thumbs.
+ * - Keep thumb values within the bounds of the slider.
+ * - Avoid visual jank while moving the thumbs
+ */
+function fixLabelOverlap(
```

Comment:
Do you think its possible to add some tests for this function? Perhaps either unit tests or maybe a few snapshot e2e playwright tests with a couple of these scenarios. Otherwise I am a little bit worried that no one will ever dare to touch the function again in the future 😅